src/client/datascience/notebook/contentProvider.ts

```diff
@@ -39,15 +40,25 @@ export class NotebookContentProvider implements VSCNotebookContentProvider {
     public get onDidChangeNotebook() {
         return this.notebookChanged.event;
     }
+    public webviews = new Map<string, NotebookCommunication[]>();
     private notebookChanged = new EventEmitter<NotebookDocumentContentChangeEvent>();
     private readonly nativeNotebookModelsWaitingToGetReloaded = new WeakMap<INotebookModel, Deferred<void>>();
     constructor(
         @inject(INotebookStorageProvider) private readonly notebookStorage: INotebookStorageProvider,
         @inject(NotebookEditorCompatibilitySupport)
-        private readonly compatibilitySupport: NotebookEditorCompatibilitySupport
-    ) {}
-    public async resolveNotebook(_document: NotebookDocument, _webview: NotebookCommunication): Promise<void> {
-        // Later
+        private readonly compatibilitySupport: NotebookEditorCompatibilitySupport,
+        @inject(IVSCodeNotebook) readonly notebookProvider: IVSCodeNotebook
+    ) {
+        notebookProvider.onDidCloseNotebookDocument(this.onDidCloseNotebook.bind(this));
+    }
+    public async resolveNotebook(document: NotebookDocument, webview: NotebookCommunication): Promise<void> {
+        // Add webviews to list. Used for testing
+        let list = this.webviews.get(document.uri.toString());
+        if (!list) {
+            list = [];
+            this.webviews.set(document.uri.toString(), list);
+        }
+        list.push(webview);
```

Comment:
Don't we need to wait for the kernel to get initialized on the UI side?

Reply:
Hmm. Maybe. None of my tests needed to wait but that might just be a coincidence. This would have to wait for the widget manager in the UI to indicate it's loaded if we wanted to truly be safe.